src/v/kafka/client/produce_partition.h

```diff
@@ -68,25 +71,50 @@ class produce_partition {
         return _batcher.consume();
     }
 
-    bool try_consume(bool timed_out) {
-        if (_in_flight || _record_count == 0) {
-            return false;
+    ss::future<> try_consume() {
+        if (!consumer_can_run()) {
+            co_return;
         }
 
-        auto batch_record_count = _config.produce_batch_record_count();
-        auto batch_size_bytes = _config.produce_batch_size_bytes();
+        _consumer(co_await do_consume());
+        co_return;
+    }
 
-        auto threshold_met = _record_count >= batch_record_count
-                             || _size_bytes >= batch_size_bytes;
+    /// \brief Arms the timer that starts the consumer
+    ///
+    /// Will arm the timer only if the consumer can run and then sets the timer
+    /// delay depending on whether or not size thresholds have been met
+    void arm_consumer() {
+        if (!consumer_can_run()) {
+            return;
+        }
 
-        if (!timed_out && !threshold_met) {
-            _timer.cancel();
-            _timer.arm(_config.produce_batch_delay());
-            return false;
+        std::chrono::milliseconds rearm_timer_delay{0};
+        // If the threshold is met, then use a delay of 0 so the timer fires
+        // nearly immediately after this call.  Otherwise use the produce
+        // batch delay when arming the timer.
+        if (!threshold_met()) {
+            rearm_timer_delay = _config.produce_batch_delay();
         }
+        _timer.cancel();
+        _timer.arm(rearm_timer_delay);
+    }
+
+    /// \brief Validates that the size threshold has been met to trigger produce
+    inline bool threshold_met() const {
```

Comment:
nitpick: I don't think the `inline` helps you much here, it's implicitly inline. I suspect you're attempting to hint the compiler to inline the function body, there's a fair chance it will ignore you, as that's not what `inline` is for.

Reply:
Yeah good point.  I think this was muscle memory from previous work I've done.